### PR TITLE
ENH: Add confusion matrix and binary confusion matrix

### PR DESCRIPTION
**Pull request by scls19fr**
Might fix https://github.com/pandas-ml/pandas-ml/issues/59

### CODE REVIEW

File: pandas_ml/confusion_matrix/abstract.py [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+    SUM_NAME_DEFAULT, COLORBAR_TRIG,
+    Backend)
+
+
+class ConfusionMatrixAbstract(object):
+    """
+    Abstract class for confusion matrix
+
+    You shouldn't instantiate this class.
+    You might instantiate ConfusionMatrix or BinaryConfusionMatrix classes
```
Comment (sinhrks) [inline pandas_ml/confusion_matrix/abstract.py]: Isn't it possible to distinguish whether ``ConfusionMatrix`` or ``BinaryConfusionMatrix`` is preferable by count the number of unique values?

File: doc/source/conf_mat.rst [hunk 2]
```diff
@@ -0,0 +1,249 @@
+Confusion matrix
+================
+
+Import ``ConfusionMatrix``
+
+.. code-block:: python
+
+    from pandas_ml.confusion_matrix import ConfusionMatrix
+
+Define actual values (``y_actu``) and predicted values (``y_pred``)
```
Comment (sinhrks) [inline doc/source/conf_mat.rst]: Can you use ``y_true`` ash the same as ``scikit-learn``?

- http://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html

**Comment by sinhrks** [comment]
Thanks. I think ``ModelFrame.metrics.confusion_matrix`` should return ``ConfusionMatrix`` instance also. To do this, we should distinguish ``ConfusionMatrix`` or ``BinaryConfusionMatrix`` automatically.

- http://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html

**Comment by scls19fr** [comment]
We should be aware that `y_true` can have 2 unique values and `y_pred` can also have 2 unique values and not to have a `BinaryConfusionMatrix` because there is no overlap between these values.
Not sure if we should use "sets" https://docs.python.org/3/tutorial/datastructures.html#sets for this.

**Comment by scls19fr** [comment]
Counting the number of rows (and columns) of confusion matrix might be enough.

About naming classes...
maybe we should name `ConfusionMatrix` like `NotBinaryConfusionMatrix`
and have a `ConfusionMatrix` to build either `ConfusionMatrix` or `NotBinaryConfusionMatrix`.

What is your opinion ?

**Comment by sinhrks** [comment]
I think the discrimination logic can be whether all labels has less than 2 unique values either ``(0, 1)`` or ``(True, False)``. Can use ``np.unique`` and check dtype. Also, let ``ConfusionMatrixAbstract.__new__``
  return a corresponding instance may be easier to use.

For class name, how about ``BinaryConfusionMatrix`` and ``LabeledConfusionMatrix``?

**Comment by scls19fr** [comment]
`ConfusionMatrix` can build now automatically either a `BinaryConfusionMatrix` or a `LabeledConfusionMatrix` but I prefer to let you now manage how to integrate confusion matrix to metrics.

**Comment by sinhrks** [comment]
> let you now manage how to integrate ...

Sure, will do. Pls merge it once when you ready.

**Comment by scls19fr** [comment]
Ok! Squashed and merged
